contracts/golos.worker/golos.worker.test.js

```diff
@@ -0,0 +1,400 @@
+const EOSTest = require("eosio.test");
+
+const eosTest = new EOSTest();
+const appName = "app.sample";
+const tokenSymbol = "APP";
+
+const delegateAccounts = [];
+const memberAccounts = [];
+
+for (let i = 0; i < 21; i++) {
+  const code = String.fromCharCode("a".charCodeAt(0) + i);
+  delegateAccounts.push(`delegate.${code}`);
+  memberAccounts.push(`user.${code}`);
+}
+
+const tokenContractPrefix = "/opt/eosio/contracts/eosio.token/eosio.token";
+let tokenContract = null;
+let contract = null;
+
+const STATE_TSPEC_APP = 1,
+  STATE_TSPEC_CREATE = 2,
+  STATE_WORK = 3,
+  STATE_TSPEC_AUTHOR_REVIEW = 4,
+  STATE_DELEGATES_REVIEW = 5,
+  STATE_PAYMENT = 6,
+  STATE_CLOSED = 7;
+
+beforeEach(async done => {
+  console.log("init");
+  await eosTest.init();
+
+  await eosTest.newAccount(
+    ...memberAccounts,
+    ...delegateAccounts,
+    "golos.worker",
+    "eosio.token",
+    appName
+  );
+
+  console.log("update contract account permissions");
+  await eosTest.api.updateauth({
+    account: "golos.worker",
+    permission: "active",
+    parent: "owner",
+    auth: {
+      threshold: 1,
+      keys: [
+        {
+          key: "EOS6MRyAjQq8ud7hVNYcfnVPJqcVpscN5So8BhtHuGYqET5GDW5CV",
+          weight: 1
+        }
+      ],
+      accounts: [
+        {
+          permission: { actor: "golos.worker", permission: "eosio.code" },
+          weight: 1
+        }
+      ]
+    }
+  });
+
+  console.log("deploy a token contract");
+  tokenContract = await eosTest.deploy(
+    "eosio.token",
+    `${tokenContractPrefix}.wasm`,
+    `${tokenContractPrefix}.abi`
+  );
+
+  console.log("build contract");
+  await eosTest.make(".");
+
+  console.log("create app token");
+  await tokenContract.create(appName, `1000000000 ${tokenSymbol}`, {
+    authorization: "eosio.token"
+  });
+
+  console.log("issue tokens");
+  for (account of [appName, ...memberAccounts]) {
+    await tokenContract.issue(
+      account,
+      `1000000 ${tokenSymbol}`,
+      "initial supply",
+      {
+        authorization: appName
+      }
+    );
+  }
+
+  console.log("deploy golos.worker contract");
+  contract = await eosTest.deploy(
+    "golos.worker",
+    "golos.worker.wasm",
+    "golos.worker.abi"
+  );
+
+  done();
+}, 300000);
+
+async function dumpState() {
+  console.log(
+    "proposals:",
+    JSON.stringify(
+      await eosTest.api.getTableRows(
+        true,
+        "golos.worker",
+        appName,
+        "proposals"
+      ),
+      null,
+      4
+    ),
+    "funds:",
+    JSON.stringify(
+      await eosTest.api.getTableRows(true, "golos.worker", appName, "funds"),
+      null,
+      4
+    )
+  );
+}
+
+async function getProposal(proposalId) {
+  return (await eosTest.api.getTableRows(
+    { json: true,
+      code: "golos.worker",
+      scope: appName,
+      table: "proposals",
+      lower_bound: proposalId,
+      limit: 1
+    })).rows[0];
+}
+
+it(
+  "1st use case test",
```

Comment:
This test checks only "positive" scenario, when all actions and parameters are right. This is not enough to be sure contract works as we expect and throws where we expect it to throw.

Please add separate (unit) tests for each action, and check following:
1. throw of each assert of an action (+check message to ensure that expected assert throws).
2. precise values check: if assert checks `< x`, than we must check boundary: `x-1` succeed, `x` fails.
3. same on timing. if something changed at arbitrary time point, we should check it *one block before* that point, then `produce_blocks(1)` and check, that state changed as expected and when expected.
4. check auths: if action pushed by wrong account, it must fail.
5. check order: some actions must fail if called in wrong state.

Reply:
Point me to the unit test example in your project.

Reply:
https://github.com/GolosChain/golos-smart/blob/master/tests/golos.ctrl_tests.cpp

Reply:
> This test checks only "positive" scenario, when all actions and parameters are right. This is not enough to be sure contract works as we expect and throws where we expect it to throw.

Should I add a functional test for the negative scripts?

Reply:
> Should I add a functional test for the negative scripts?

It's not necessary, unit-tests will cover it